Revised firmware calibration and upgrade checks

wavexx opened the pull request:
In order to *properly* handle new optional calibration steps such as temperature model calibration the entire wizard/calibration workflow had to be revised. A new eeprom variable has been introduced that handles the calibration stages as a bitfield (fixing #3836) and cleans up most of the weird calibration handling of the previous versions. The variable `EEPROM_CALIBRATION_STATUS_V2` properly handles upgrade workflows and new calibration features transparently. The old variable (now `EEPROM_CALIBRATION_STATUS_V1`) is kept unused to allow downgrading to a previous version.

With this change we can now explicitly perform temperature model calibration after a firmware upgrade, so that the default model values are not used on existing printers which already completed the full wizard workflow. We currently do not ask: we just warn that the temperature model will be performed. This is similar in behavior to FW 3.2 where selftest was enforced to perform belt calibration.

Due to this change there are a lot of things that needs to be retested surrounding the wizard handling:

- ALTFAN override initialization has been cleaned up. Ensure ALTFAN is still detected and handled properly when installed after a FACTORY RESET.
- Ensure ALTFAN behavior is not changed if FW is upgraded *without* a FW reset.
- Ensure the entire wizard works in the same order as before.
- Ensure the wizard still asks about calibrating more sheets as before.
- Ensure the wizard after a factory prep reset works as before.
- When upgrading from FW 3.2.0.3 (or older) self-test needs to be enforced as before.
- When upgrading from FW 3.11 (or older) there is now a new message warning that the model calibration will start.
- If the full wizard is skipped, the usual "printer is not calibrated, follow calibration flow" should be shown.
- If the wizard is aborted due to selftest failure, hotend or xyz or temperature model calibration failure there should be a warning about incomplete wizard. This wasn't handled properly before.
- Ensure the wizard still terminates properly on failure (due to selftest, hotend, xyz or model failure) and doesn't attempt to continue indefinitely.
- Even if the wizard is skipped due to a later failure, a warning about a missing temperature model calibration should be shown. The warning is NOT shown if the wizard is skipped intentionally at the first question.
- The warning should NOT be shown if the model is disabled afterwards.
- You should be able to RESET the printer to abort temp model calibration (the next boot should show that model calibration is required, but it not done automatically again).
- Even if the wizard is skipped due to failure, a warning about a missing Z offset should be shown (as before).
- The same warning should be shown when switching to a new (unused) bed sheet.
- The warning should go away if we switch back to a known sheet and reset.
- Ensure we can downgrade to a previous version without failure.

Doing a "SHIPPING PREP" reset should require (and do) _only_ Z calibration after power-up. This assumes model calibration will be done in the factory. We can adjust this if needed and it's easy to do so now.

There is a new string and an updated string regarding the temperature model calibration to be updated.

PFW-1469

wavexx committed:
Simplify FW eeprom version checks

- Generalize force_selftest_if_fw_version() so we can check when upgrading
  between arbitrary FW versions.
- Do not call update_current_firmware_version_to_eeprom() prematurely
  when forcing self-test (if a reset happens before self-test completes,
  the check would be incorrectly skipped on the next run).

wavexx committed:
eeprom: Fix EEPROM_HEAT_BED_ON_LOAD_FILAMENT debug offset in docs

wavexx committed:
Replace calibration state with a new calibration bitmap

- Move calibration status/declarations to util.h
- Allows to query the status of each step independently without assuming
  any order
- Allows to extend the calibration with new steps transparently

wavexx committed:
Simplify altfan override initialization

wavexx committed:
wizard: Remove unused wizard stages

Remove PreheatPla/IsPla, as we currently always explicitly ask for the
material type.

wavexx committed:
Handle CALIBRATION_STATUS_LIVE_ADJUST

wavexx committed:
Handle CALIBRATION_STATUS_{Z,XYZ}

wavexx committed:
Handle CALIBRATION_STATUS_SELFTEST

wavexx committed:
Handle CALIBRATION_STATUS_TEMP_MODEL

wavexx committed:
Unbreak build with !PINDA_THERMISTOR

wavexx committed:
wizard: Rewrite the wizard handling loop

Simplify status tracking:

- S::Restore to continue to the next logical wizard item
- S::Finish for a successful failure
- S::Failed to exit while showing a failure

wavexx committed:
wizard: Rewrite the fw upgrade check logic

- Handle the conversion from the V1 calibration status to the V2 bitmap
  for all past FW versions.
- Allow to run the wizard in "Restore" mode (for example during upgrade
  or service prep) to complete any missing steps.
- Fix service prep to use the above feature by clearing the appropriate
  calibration bits.
- Remove exceptions for FW 3.2.0.4 which can now be handled cleanly.

wavexx commented:
There is currently no way to set model as "calibrated". For example, when setting all values via M310: the printer will still moan that model calibration has not been performed.

I do not want to abuse existing flags in M310 for this.

D10 allows to set the XYZ calibration status, I wonder if we need the same for M310.

wavexx commented:
Also note: when applying this to 3.12, we should still keep the same eeprom offset even though EEPROM_HEAT_BED_ON_LOAD_FILAMENT is not available.

3d-gussner commented:
> Also note: when applying this to 3.12, we should still keep the same eeprom offset even though EEPROM_HEAT_BED_ON_LOAD_FILAMENT is not available.

Easiest is then also cherry-pick that PR to MK3_3.12 so we have the that in sync, or not?

3d-gussner commented:
First tests are working well. I want to run some additional tests with serial debug of the calibration status.

DRracer commented on Firmware/util.cpp:
```diff
@@ -181,32 +181,23 @@ inline int8_t is_provided_version_newer(const char *version_string)
 }
 
-bool force_selftest_if_fw_version()
+bool eeprom_fw_version_older_than(const uint16_t (&ver_req)[4])
```
we can probably optimize the parameter as PROGMEM

wavexx replied:
Done

DRracer commented on Firmware/util.cpp:
```diff
@@ -522,3 +513,24 @@ void ip4_to_str(char* dest, uint8_t* IP)
 }
+
+
+bool calibration_status_get(CalibrationStatus components)
```
Are these functions inlined? Does it make sense making them __attribute__((noiniline))?

wavexx replied:
I was /about/ to say they _cannot_ be, but with -flto you can't say right away.
They're not inlined. As they're all external, they should never be.

DRracer reviewed (approved):
very nice, generally I approve this PR :+1: 

there are a few minor comments which can be addressed later

wavexx committed:
wizard: Allow partial resuming after reset

Update the eeprom FW version as soon as migration is complete, to avoid
resetting the CALIBRATION_V2 variable at each reset.

Do not implicitly reset the calibration steps for WizState::Run: do this
only for the menu action.

wavexx committed:
wizard: Disable temperature model during hotend testing

This avoids confusing error messages if the hotend is defective and
caught by selftest itself.

Rearrange Stopped state checks (saves 2 bytes)

wavexx committed:
Read version data from progmem in eeprom_fw_version_older_than

wavexx committed:
wizard: Resume the wizard for missing calibrations during upgrades

Instead of checking for explicit versions, resume the wizard if some
(new) wizard is missing.

This handles both the old SELFTEST check, the new thermal model
and any future check in the same fashion.

wavexx committed:
wizard: Ask about the sheet before doing Z calibration

Ask to install the sheet immediately after removing the test print.
This makes further messages less confusing.

3d-gussner commented:
Tested on MK404 and it works well, just one or two issue I have found.
- Factory reset ALL :heavy_check_mark: 
- Factory reset Shipping/Service :heavy_check_mark:
- LCD Selftest :heavy_check_mark:
- LCD TM cal :heavy_check_mark:
- XYZ cal. :question: As it resets the cal status from 31 to 15 and not back to 31 after finished
- Same for MMU2SR :heavy_check_mark:
- No thermal anomaly during hotend selftest :heavy_check_mark:
- Reset the printer during Wizard it continues (if selected with yes) at the missed wizard step :heavy_check_mark: 
- Minor issue `calibration_status_clear` and maybe `calibration_status_set` update the eeprom value even it didn't change.
